File: typescript/docs/snippets/charge.client.ts
// Client-side charge: pay-kit settles the 402 over MPP or x402, then retries.
// See ./README.md for the snippet:start/end convention.

import { createPayKitClient } from '@solana/pay-kit/client'
import { createKeyPairSignerFromBytes, getBase58Encoder } from '@solana/kit'

declare const SECRET: string

async function main(): Promise<void> {
  // snippet:start
  const signer = await createKeyPairSignerFromBytes(getBase58Encoder().encode(SECRET))
  const client = await createPayKitClient({ signer, rpcUrl: 'http://localhost:8899' })

  // On a 402, pay-kit pays with the matching protocol and retries — transparently.
  const res = await client.fetch('${URL}')
  console.log(await res.json())
  // snippet:end
}

void main

File: rust/docs/snippets/charge.client.rs
//! Client-side charge: read the 402 challenge, sign a credential, retry.
//!
//! Mirrors the Client section of `rust/README.md`. See
//! `../../../docs/snippets-convention.md` for the snippet:start/end convention.

use solana_pay_kit::mpp::client::{build_credential_header, parse_challenge};
use solana_pay_kit::solana_keychain::memory::MemorySigner;
use solana_pay_kit::mpp::solana_rpc_client::rpc_client::RpcClient;

async fn pay(
    www_authenticate: &str,
    signer: &MemorySigner,
    rpc: &RpcClient,
) -> Result<(), Box<dyn std::error::Error>> {
    // snippet:start
    // 1. Parse the MPP charge challenge from the 402 `WWW-Authenticate` header.
    let challenge = parse_challenge(www_authenticate)?;

    // 2. Sign a payment credential and replay the request with it.
    let authorization = build_credential_header(signer, rpc, &challenge).await?;
    let res = reqwest::Client::new()
        .get("${URL}")
        .header("Authorization", authorization)
        .send()
        .await?;
    println!("{}", res.text().await?);
    // snippet:end
    Ok(())
}
